src/devices/PiJuice/PiJuice.cs

```diff
@@ -0,0 +1,240 @@
+        public void SetIdEepromAddress(IdEepromAddress epromAddress)
+        {
+            WriteCommandVerify(PiJuiceCommand.IdEepromAddress, new byte[] { (byte)epromAddress });
+        }
+
+        /// <summary>
+        /// Write a PiJuice command
+        /// </summary>
+        /// <param name="command">The PiJuice command</param>
+        /// <param name="data">The data to write</param>
+        internal void WriteCommand(PiJuiceCommand command, byte[] data)
+        {
+            byte tries = 0;
+            IOException innerEx = null;
+            // When writing/reading to the I2C port, PiJuice doesn't respond on time in some cases
+            // So we wait a little bit before retrying
+            // In most cases, the I2C read/write can go thru without waiting
+            while (tries < MaxRetries)
+            {
+                try
+                {
+                    // create request
+                    var buffer = new byte[data.Length + 2];
+                    Array.Copy(data, 0, buffer, 1, data.Length);
+                    buffer[0] = (byte)command;
+                    buffer[buffer.Length - 1] = GetCheckSum(data, all: true);
+                    _i2cDevice.Write(buffer);
+                    return;
+                }
+                catch (IOException ex)
+                {
+                    // Give it another try
+                    innerEx = ex;
+                    tries++;
+                    DelayHelper.DelayMilliseconds(ReadRetryDelay, false);
+                }
+            }
+
+            throw new IOException($"{nameof(WriteCommand)}: Failed to write command {command}", innerEx);
+        }
+
+        /// <summary>
+        /// Write a PiJuice command and verify
+        /// </summary>
+        /// <param name="command">The PiJuice command</param>
+        /// <param name="data">The data to write</param>
+        /// <param name="delay">The delay before reading the data</param>
+        internal void WriteCommandVerify(PiJuiceCommand command, byte[] data, int delay = 0)
+        {
+            WriteCommand(command, data);
+
+            if (delay > 0)
+            {
+                DelayHelper.DelayMilliseconds(delay, false);
+            }
+
+            var response = ReadCommand(command, (byte)data.Length);
+
+            if (!data.SequenceEqual(response))
+            {
+                throw new IOException($"{nameof(WriteCommandVerify)}: Failed to write command {command}");
+            }
+        }
+
+        /// <summary>
+        /// Read data from PiJuice
+        /// </summary>
+        /// <param name="command">The PiJuice command</param>
+        /// <param name="length">The length of the data to be read</param>
+        /// <returns>Returns an array of the bytes read</returns>
+        internal byte[] ReadCommand(PiJuiceCommand command, byte length)
+        {
+            byte[] outArray = new byte[length + 1];
```

Comment:
This should be a Span<byte> with stackalloc. You're using it with I2cDevice.Read. So better use a Span<byte>

Reply:
Should the ReadCommand still return a byte array or should it return a Span<byte> ?

Reply:
It's clear that something like `void ReadCommand(PiJuiceCommand command, Span<byte> readBuffer)` would be a better design. Same for the Write command. In this case, you would have to adjust the readBuffer size to match the length+1 size. Similar for the write command. Now those are internals, I saw how it's used for the rest, you just grab one or 2 bytes every time. I don't have a super strong opinion about it.
@krwq may have a stronger opinion.

Reply:
As an example then

```csharp
WriteCommand(PiJuiceCommand.ResetToDefault, new byte[] { 0xAA, 0x55, 0x0A, 0xA3 });
   ```

would become

```csharp
Span<byte> buffer = stackalloc byte[4 + 2]
{
    0,
    0xAA,
    0x55,
    0x0A,
    0xA3,
    0
};

WriteCommand(PiJuiceCommand.ResetToDefault, buffer);
   ```

Reply:
For the read command

```csharp
var response = ReadCommand(PiJuiceCommand.FirmwareVersion, 2);
```

Would become

```csharp
Span<byte> buffer = stackalloc byte[2 + 1];
buffer.Clear();

ReadCommand(PiJuiceCommand.FirmwareVersion, buffer);

var response = buffer.Slice(0, 2);
```

You would need to slice the buffer in the calling method as this is currently done in the ReadCommand

Reply:
As I said, I have no strong opinion on this. I always design API with Span<byte> when it's public. When private sometimes, it's with byte[] or Span<byte> when it does make sense.

Reply:
I will leave it as is.